core/iwasm/common/wasm_c_api.c

```diff
@@ -4711,65 +4650,108 @@ aot_process_export(wasm_store_t *store, const AOTModuleInstance *inst_aot,
 }
 #endif /* WASM_ENABLE_AOT */
 
-wasm_instance_t *
-wasm_instance_new(wasm_store_t *store, const wasm_module_t *module,
-                  const wasm_extern_vec_t *imports, own wasm_trap_t **trap)
-{
-    return wasm_instance_new_with_args(store, module, imports, trap,
-                                       KILOBYTE(32), KILOBYTE(32));
-}
-
+/**
+ * Assume `imports` follows the order of "FUNCITONS, GLOBALS, MEMORIES and
+ * TABLES" which is the order of `out` of `wasm_module_import`
+ */
 static bool
-compare_imports(const wasm_module_t *module, const wasm_extern_vec_t *imports)
+do_link(const wasm_instance_t *inst, const wasm_module_t *module,
+        const wasm_extern_vec_t *imports)
 {
-    unsigned import_func_count = 0;
-    unsigned import_global_count = 0;
-    unsigned import_memory_count = 0;
-    unsigned import_table_count = 0;
-    unsigned i = 0;
+    uint32 i = 0, import_func_i, import_global_i;
+    uint32 import_func_count = 0, import_global_count = 0;
+
+    bh_assert(inst && module);
+
+#if WASM_ENABLE_INTERP != 0
+    if ((*module)->module_type == Wasm_Module_Bytecode) {
+        import_func_count = MODULE_INTERP(module)->import_function_count;
+        import_global_count = MODULE_INTERP(module)->import_global_count;
+    }
+#endif
+#if WASM_ENABLE_AOT != 0
+    if ((*module)->module_type == Wasm_Module_AoT) {
+        import_func_count = MODULE_AOT(module)->import_func_count;
+        import_global_count = MODULE_AOT(module)->import_global_count;
+    }
+#endif
 
-    for (i = 0; imports && i < imports->num_elems; i++) {
+    /* we have run a module_type check before. */
+
+    /* 1. imported functions */
+    for (i = 0, import_func_i = 0;
+         i < imports->num_elems && import_func_i < import_func_count;
+         i++, import_func_i++) {
```

Comment:
Must the import types of functions and globals in the import list of wasm_instance be ordered? e.g. first goud of functions, and then the group of globals? Could globals be before functions, or mixed of them with ordered functions and ordered globals?

Reply:
The new recommendation is different with before.  

Before, we required developers to make a list of import and keep them in the right order. But now, we only ask developers to prepare all necessary *imports* and follow the order information providing by `wasm_module_imports()`. That is what `build_imports()` does.